src/Infrastructure/LeanCode.Firebase.FCM/NotificationDataConverter.cs

```diff
@@ -0,0 +1,58 @@
+using System.Collections.Immutable;
+using System.Reflection;
+
+namespace LeanCode.Firebase.FCM;
+
+public sealed class NotificationDataConverter
+{
+    private const string TypeField = "Type";
+
+    private ImmutableDictionary<Type, Func<object, string>> formatters = ImmutableDictionary<
+        Type,
+        Func<object, string>
+    >.Empty;
+
+    public void FormatUsing<T>(Func<T, string> formatter)
+    {
+        formatters = formatters.Add(typeof(T), o => formatter((T)o));
+    }
+
+    /// <summary>
+    /// Converts POCO object to a notification data dictionary. Does not support hierarchical
+    /// data.
+    /// </summary>
+    public Dictionary<string, string> ToNotificationData(object data)
```

Comment:
Random question and unrelated to problem at hand but why isn't this generic?

```suggestion
    public Dictionary<string, string> ToNotificationData<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)] T>(T data)
```

Reply:
What will we gain by making it generic?

Reply:
Avoids boxing with value types, one less reflection to fix before AOT complilation works, etc.

Not that they would be some huge benefits but I'm just curious why was the initial decision made to force conversion to `object`. What did we gain by not making it generic?

Reply:
Any value type will be boxed anyway with `prop.GetValue(data)` call. Btw, this change will make boxing happen per every property, won't it? 
I see no generic gains here.

Reply:
> Any value type will be boxed anyway with `prop.GetValue(data)` call.

Ehh, you're right, I totally forgot about this.